### PR TITLE
Synchronized TorchModelV2 for successfully running PyTorch with AsyncSampler

### PR DESCRIPTION
**Pull request by roireshef**
## Why are these changes needed?

This introduces thread safety into TorchModelV2 for decentralized gradient computing algorithms, such as A3C. This is crucial when using pytorch with `async_sampler=True`, which achieves ~25%-50% improvement (depending on `num_envs_per_worker`) in sampling bandwidth over the default `async_sampler=False` option 

## Related issue number

Implements proposal (1) in https://github.com/ray-project/ray/issues/10628 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by roireshef** [comment]
Sorry guys, I wasn't able to run the tests nor lint, as my setup is partial and based on `scripts/setup-dev.py`

**Comment by roireshef** [comment]
Seems like the Travis job has a broken test in which np.float32 is found not to be of "floating" type.

File: rllib/models/modelv2.py [hunk 1]
```diff
@@ -229,6 +228,16 @@ def __call__(
         self._last_output = outputs
         return outputs, state
 
+    def call_with_value(
+            self,
+            input_dict: Dict[str, TensorType],
+            state: List[Any] = None,
+            seq_lens: TensorType = None) -> (TensorType, TensorType, List[TensorType]):
+        """ Convenience function to use where both model out and value function are required """
```
Comment (sven1977) [inline rllib/models/modelv2.py:236]:
```suggestion
        """Use this if both model's output and -value are required (thread safe)."""
```

File: rllib/models/modelv2.py [hunk 2]
```diff
@@ -248,6 +257,27 @@ def from_batch(self, train_batch: SampleBatch,
         del train_batch["is_training"]
         return ret
 
+    @PublicAPI
+    def from_batch_with_value(self, train_batch: SampleBatch,
+                              is_training: bool = True) -> (TensorType, TensorType, List[TensorType]):
+        """Convenience function that calls this model with a tensor batch and calls value function atomically so that
+            competing threads aren't exposed to other thread's value_function outputs
```
Comment (sven1977) [inline rllib/models/modelv2.py:264]:
```suggestion
        """Calls model and value function in thread-safe manner."""
```

File: rllib/models/utils.py [hunk 3]
```diff
@@ -65,3 +67,16 @@ def get_initializer(name, framework="tf"):
 
     raise ValueError("Unknown activation ({}) for framework={}!".format(
         name, framework))
+
+
+class Synchronizable:
+    """ Interface for classes that require instance-level thread locking for their methods """
```
Comment (sven1977) [inline rllib/models/utils.py:73]:
```suggestion
    """Interface for classes that require instance-level locking for their methods.
    """
```

File: rllib/models/utils.py [hunk 4]
```diff
@@ -65,3 +67,16 @@ def get_initializer(name, framework="tf"):
...
+    def __init__(self):
+        self.synchronized_lock = threading.RLock()
+
+    @staticmethod
+    @decorator
+    def synchronized(wrapped, instance=None, args=None, kwargs=None):
+        """ wrapper used to decorate the instance methods that should be locked with self.synchronized_lock """
```
Comment (sven1977) [inline rllib/models/utils.py:80]:
```suggestion
        """Decorator for instance methods that should be locked (thread safe).
        """
```

**Review by sven1977 — APPROVED** [review]
Thanks @roireshef Looks great, just fixed some LINT errors and waiting for tests to pass.
Also, could you pull from master again to make the conflict go away?

**Comment by sven1977** [comment]
Took care already of the conflict. Just waiting for tests again now (just in case) ...

**Comment by sven1977** [comment]
Win tests are quite flaky at the moment. We are working on getting those more stable.

File: rllib/models/modelv2.py [hunk 5]
```diff
@@ -250,6 +259,26 @@ def from_batch(self, train_batch: SampleBatch,
         ret = self.__call__(input_dict, states, input_dict.get("seq_lens"))
         return ret
 
+    @PublicAPI
+    def from_batch_with_value(self, train_batch: SampleBatch,
+                              is_training: bool = True) -> (TensorType, TensorType, List[TensorType]):
+        """Calls model and value function in thread-safe manner."""
```
Comment (sven1977) [inline rllib/models/modelv2.py:265]:
```suggestion
        """Calls model and its value function in a thread-safe manner.
```

File: rllib/models/modelv2.py [hunk 6]
```diff
@@ -231,6 +230,16 @@ def __call__(
         self._last_output = outputs
         return outputs, state
 
+    def call_with_value(
+            self,
+            input_dict: Dict[str, TensorType],
+            state: List[Any] = None,
+            seq_lens: TensorType = None) -> (TensorType, TensorType, List[TensorType]):
+        """Use this if both model's output and -value are required (thread safe)."""
```
Comment (sven1977) [inline rllib/models/modelv2.py:238]:
```suggestion
        """Used if both model's output and -value are required (thread safe).
        """
```

File: rllib/models/utils.py [hunk 7]
```diff
@@ -142,3 +144,17 @@ def get_initializer(name, framework="tf"):
...
+class Synchronizable:
+    """Interface for classes requiring instance-level locking for methods.
+    """
+    def __init__(self):
+        self.synchronized_lock = threading.RLock()
+
+    @staticmethod
+    @decorator
+    def synchronized(wrapped, instance=None, args=None, kwargs=None):
+        """Decorator for instance methods that should be locked (thread safe)."""
```
Comment (sven1977) [inline rllib/models/utils.py:158]:
```suggestion
        """Decorator for instance methods that should be locked (thread safe).
        """
```

File: rllib/models/modelv2.py [hunk 8]
```diff
@@ -250,6 +260,26 @@ def from_batch(self, train_batch: SampleBatch,
...
+        right input dict, state, and seq len arguments.
+        """
+
+        train_batch["is_training"] = is_training
+        states = []
+        i = 0
+        while "state_in_{}".format(i) in train_batch:
+            states.append(train_batch["state_in_{}".format(i)])
+            i += 1
+        ret = self.call_with_value(train_batch, states, train_batch.get("seq_lens"))
```
Comment (sven1977) [inline rllib/models/modelv2.py:278]:
```suggestion
        ret = self.call_with_value(train_batch, states,
                                   train_batch.get("seq_lens"))
```

File: rllib/models/utils.py [hunk 9]
```diff
@@ -142,3 +144,18 @@ def get_initializer(name, framework="tf"):
 
     raise ValueError("Unknown activation ({}) for framework={}!".format(
         name, framework))
+
+
+class Synchronizable:
+    """Interface for classes requiring instance-level locking for methods.
+    """
```
Comment (sven1977) [inline rllib/models/utils.py:151]:
```suggestion
    """

```

File: rllib/agents/a3c/a3c_torch_policy.py [hunk 10]
```diff
@@ -80,8 +79,8 @@ def torch_optimizer(policy, config):
 
 class ValueNetworkMixin:
     def _value(self, obs):
-        _ = self.model({"obs": torch.Tensor([obs]).to(self.device)}, [], [1])
-        return self.model.value_function()[0]
+        _, value, _ = self.model.call_with_value({"obs": torch.Tensor([obs]).to(self.device)}, [], [1])
```
Comment (sven1977) [inline rllib/agents/a3c/a3c_torch_policy.py:82]:
```suggestion
        _, value, _ = self.model.call_with_value({
            "obs": torch.Tensor([obs]).to(self.device)
        }, [], [1])
```

File: rllib/models/modelv2.py [hunk 11]
```diff
@@ -231,6 +230,17 @@ def __call__(
         self._last_output = outputs
         return outputs, state
 
+    def call_with_value(
+            self,
+            input_dict: Dict[str, TensorType],
+            state: List[Any] = None,
+            seq_lens: TensorType = None) -> (TensorType, TensorType, List[TensorType]):
```
Comment (sven1977) [inline rllib/models/modelv2.py:237]:
```suggestion
    def call_with_value(self,
                        input_dict: Dict[str, TensorType],
                        state: List[Any] = None,
                        seq_lens: TensorType = None
                        ) -> (TensorType, TensorType, List[TensorType]):
```

File: rllib/models/modelv2.py [hunk 12]
```diff
@@ -250,6 +260,26 @@ def from_batch(self, train_batch: SampleBatch,
         ret = self.__call__(input_dict, states, input_dict.get("seq_lens"))
         return ret
 
+    @PublicAPI
+    def from_batch_with_value(self, train_batch: SampleBatch,
+                              is_training: bool = True) -> (TensorType, TensorType, List[TensorType]):
```
Comment (sven1977) [inline rllib/models/modelv2.py:265]:
```suggestion
    def from_batch_with_value(self,
                              train_batch: SampleBatch,
                              is_training: bool = True
                              ) -> (TensorType, TensorType, List[TensorType]):
```

File: rllib/models/modelv2.py [hunk 13]
```diff
@@ -33,8 +33,7 @@ class ModelV2:
                value_function() -> V(s)
     """
 
-    def __init__(self, obs_space: gym.spaces.Space,
-                 action_space: gym.spaces.Space, num_outputs: int,
+    def __init__(self, obs_space: gym.spaces.Space, action_space: gym.spaces.Space, num_outputs: int,
```
Comment (sven1977) [inline rllib/models/modelv2.py:36]:
```suggestion
    def __init__(self, obs_space: gym.spaces.Space,
                 action_space: gym.spaces.Space, num_outputs: int,
```

**Comment by roireshef** [comment]
@sven1977 thanks for taking care of this PR. do you need any help from my end here?

**Comment by sven1977** [comment]
Hey @roireshef we did merge another slightly more light-weight solution recently (which did not require a new dependency: wrapt). PR: https://github.com/ray-project/ray/pull/13467

**Pull request closed without merging**